### DISCUSSION TITLE
Splitting an app into modules

### DISCUSSION THREAD

**Discussion by CraigSiemens** [MobileNativeFoundation/discussions#72] [General]
Generally as apps grow, there's some benefits that can come from having a more modular app. I'm curious to hear the approach other apps have taken to split up into multiple modules. I'd love any answers others might have to some of the following questions.

1. What are some of the pros and cons you've encounter with multiple modules?
2. Do you keep them in the same repo, or do you split them out multiple?
3. If you split them out, do you version them?
4. Do you have guidelines you follow around when to create a new module, or is it just as needed?
5. Do you have guidelines for the "kinds" of modules, or just just try to keep related things together?
6. Have you made any changes to your build system as a result of using multiple modules?
7. Any other related tooling you've added?

---

At Neo Financial, the iOS app was started with multiple modules. We started with some core modules for reusable UI and common extensions and utilities. We've been gradually splitting the app into more modules as its grown, generally split by distinct flows in the UI like some onboarding flows or the tabs in the app. Now we're at 26 modules.

1. 
- Pros
  - it forces us to think more about where things should live and what other places will need to use it.
  - Can give faster build times when working on a lower level modules. We've got separate schemes in our Xcode project for each one.
- Cons
  - it's more management, periodically taking one module and splitting it up into multiple
  - can also give slower build times when changing something at a lower level then building a higher level module. Multiple models that depend on it can end up building as well.
2. It's all in a single repo.
3. N/A
4. Unfortunately, not really. We usually make new ones when we notice that two places of the app now need to share a flow/assets/logic and it doesn't make sense for one to depend on the other.
5.
- Modules that contain a flow in the UI. They generally contain a `Coordaintor` that's the entry point for another module to use it. They also contain some or all of the models and API requests.
- Core modules that contain the stuff that obvious needs to be shared. We've got ones for the base API requests and user session management, UI, and extension/utilities.
- Wrappers around dependencies. We use these to abstract away things like analytics destinations, remove config values, and some are by necessity for things like google maps that's distributed as a static framework and would be duplicated if imported into multiple other dynamic frameworks.
- Other code related to a certain domain. For example our `User` module (I wish it had a better name) contains shared stuff that both in the onboarding flows, and later in the user setting screens for things like entering home address.
6. Nope, we're just using Xcode to build the targets.
7. We use XcodeGen for creating the Xcode project based on the folder structure and some yml files. We also have a CLI tool that can automate some tasks for creating or updating a new module. It simplifies some of the setup for tests, using SwiftGen for assets and strings,  and Apollo for API request codegen. XcodeGen also allows us to programmatically generate some `xcfilelists` daily easily that are used in different build phases so they can skip running if nothing has changed in the related files.

**Comment by hendych** [comment]
@CraigSiemens Nice share.  In Traveloka, we have product 100+ modules. 

- We started by drawing our module dependencies and defining module principles to prevent cyclic dependency.
- We start from common modules (which owned by our central team) because all modules depend on common modules.
- Product team has their own code module, where they store public objects. They start from their core modules and then their product modules (feature modules). Each product modules interact with another product modules with `Coordinator` so they don't have to depend directly.
- We are using modified CocoaPods template `pod-template` to generate the module template and manage dependencies between modules.
- We created SandboxApp in product modules, which is a micro app, so developers can develop and build their features (UI) faster.
- We created a module that manages 3rd party initializations so we can keep environments between SandboxApp and our main project app as similar as possible.

Fore more details, check out our journey [here](https://medium.com/traveloka-engineering/our-journey-in-reducing-traveloka-ios-apps-build-time-by-90-3f875ff9a9b7)!

**Comment by r00tify** [comment]
Hi Craig,

Thank you for starting this discussion. While I don't have a lot to share about this topic, I wanted to add a few other question that I would love to hear the community's opinion about:

8. How do you manage the modules? Do you just use .xcodeproj files in a workspace or you manage them with a dependency manager?
9. How do you link the modules (statically or dynamically)?
10. If you link the modules statically, how do you prevent frameworks duplication if a framework is used in multiple other frameworks?
11. If you link the modules dynamically, how do you keep launch time reasonable with multiple dylibs?

**Reply by keith** [reply]
For us:

8. We use bazel but before that we were using xcodegen which worked great for this
9. We link statically
10. We only have a few very small extensions and no other dynamic binaries so we allow the duplication of some of our core libraries
11. We don't hit this but supposedly this is mostly fixed by changes to dyld, I'd love to hear from folks who have tested this

**Reply by donnie-jp** [reply]
In our team's case:
8. We deploy our modules (SDKs) to CocoaPods - the open sourced ones on trunk and the internal ones on a private spec repo. We are now starting to add SPM support too.
9. It's up to the app - most apps use dynamic linking and some use static.
10. Not hit this.
11. Yeah this is a concern.

**Reply by CraigSiemens** [reply]
8. We use XcodeGen to create one xcodeproj that contains all the models in it.
9. Dynamically
10. n/a
11. Our app still launches faster than the animation to show it so it hasn't been a concern yet.

**Reply by hendych** [reply]
Answering your question:
8. We used CocoaPods as our dependency management, using monorepo so we don't publish the internal modules, instead we used `:path =>` in CocoaPods. We also used Xcodegen to prevent conflicts in the legacy (monolith app) project. After we used Buck, it seems that we don't need dependency management again for our internal module.
9. Statically. We have 100+ modules, this will be an issue if we linked it dynamically.
10. If you linked all frameworks statically, this should be not an issue. Unless you tried to link the static framework into a dynamic one, and then tried to link the dynamic framework again into the static one. 
11. Don't use too many dynamic frameworks. If you do, try not to load the dynamic framework on launch time. I also read somewhere that you can also merge the modules into 1 dynamic framework. Instead of 1 module 1 dynamic framework, you can merge 1 dynamic framework for example 2 modules or more. This can be achieved easily if you use CocoaPods. https://github.com/grab/cocoapods-pod-merge. There are many articles on how to solve this. WWDC apple https://developer.apple.com/documentation/xcode/reducing-your-app-s-launch-time/

**Comment by shepting** [comment]
Hi Craig! 👋 

We are currently at 1300+ modules for the Airbnb app. Part of the motivation for the large number of modules is to create "Interface Modules" which is actually what other modules will depend upon to improve build times and allow for better parallelization of our Swift builds. It means that fewer modules will need to be rebuilt if only code _inside_ a feature module has been changes (less cascading rebuilds).

**Reply by ileitch** [reply]
Hey @shepting,

I'm curious if you any thoughts/plans to alter this approach given the new incremental build behavior coming in Swift 5.5, with regards to modules? Explained here beginning at 12:50 https://developer.apple.com/videos/play/wwdc2021/10192/?time=768

**Reply by dive** [reply]
Hi @shepting,

Thank you for sharing! Can you also share your experience with Xcode and 1300+ modules? Do you use Bazel or the default Xcode build system? Is it possible to open a workspace with 1300+ modules at all? 😅

**Reply by JohnnyTseng** [reply]
Hi @shepting , 

Does these all 1300+ modules (frameworks) link to Airbnb iOS app? 
Does it affect the launch time & binary size of the App? If so, how do you cope with this issue? 

Thank you!

**Reply by keith** [reply]
> Hey @shepting,
> 
> I'm curious if you any thoughts/plans to alter this approach given the new incremental build behavior coming in Swift 5.5, with regards to modules? Explained here beginning at 12:50 [developer.apple.com/videos/play/wwdc2021/10192/?time=768](https://developer.apple.com/videos/play/wwdc2021/10192/?time=768)

I'm not Steven but we're super excited about this, it should solve a major problem with modularized projects by stopping the builds from cascading if there aren't any changes requiring modules up the dependency tree to be recompiled.



> Hi @shepting,
> 
> Thank you for sharing! Can you also share your experience with Xcode and 1300+ modules? Do you use Bazel or the default Xcode build system? Is it possible to open a workspace with 1300+ modules at all? 😅

Lyft has a similar number of modules, around 1800 source modules, most of which have corresponding test targets, but we have a setup where we generate partial Xcode projects depending on what developers are working on, including only the targets they care about, which helps Xcode performance alot

> Does it affect the launch time & binary size of the App? If so, how do you cope with this issue?

We statically link ours so it doesn't affect this.
